libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/history/ReplayHistorySession.kt

```diff
@@ -0,0 +1,115 @@
+import com.mapbox.navigation.core.lifecycle.MapboxNavigationObserver
+import com.mapbox.navigation.utils.internal.logI
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.update
+
+/**
+ * Used to create a replay trip session.
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+class ReplayHistorySession : MapboxNavigationObserver {
+
+    private val optionsFlow = MutableStateFlow(
+        ReplayHistorySessionOptions.Builder().build()
+    )
+    private var mapboxHistoryReader: MapboxHistoryReader? = null
+    private var mapboxNavigation: MapboxNavigation? = null
+    private var lastHistoryEvent: ReplayEventBase? = null
+    private var coroutineScope: CoroutineScope? = null
+
+    private val replayEventsObserver = ReplayEventsObserver { events ->
+        logI(LOG_CATEGORY) { "replayEvents ${events.joinToString("\n")}" }
+        if (isLastEventPlayed(events)) {
+            pushMorePoints()
+        }
+    }
+
+    private fun isLastEventPlayed(events: List<ReplayEventBase>): Boolean {
+        val currentLocationEvent = events.lastOrNull { it is ReplayEventUpdateLocation }
+            ?: return false
+        val lastEventTimestamp = this.lastHistoryEvent?.eventTimestamp ?: 0.0
+        return currentLocationEvent.eventTimestamp >= lastEventTimestamp
+    }
+
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        val coroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
+            .also { this.coroutineScope = it }
+        this.mapboxNavigation = mapboxNavigation
+        this.lastHistoryEvent = null
+        mapboxNavigation.startReplayTripSession()
+        observeStateFlow(mapboxNavigation).launchIn(coroutineScope)
+        mapboxNavigation.mapboxReplayer.registerObserver(replayEventsObserver)
+    }
+
+    inline fun <T, R> Flow<T>.mapDistinct(
+        crossinline transform: suspend (value: T) -> R
+    ): Flow<R> = map(transform).distinctUntilChanged()
+
+    override fun onDetached(mapboxNavigation: MapboxNavigation) {
+        coroutineScope?.cancel()
+        coroutineScope = null
+        this.mapboxNavigation = null
+        mapboxNavigation.mapboxReplayer.unregisterObserver(replayEventsObserver)
+        mapboxNavigation.mapboxReplayer.stop()
+        mapboxNavigation.mapboxReplayer.clearEvents()
+    }
+
+    fun getOptions(): StateFlow<ReplayHistorySessionOptions> = optionsFlow.asStateFlow()
+
+    fun setOptions(options: ReplayHistorySessionOptions) = optionsFlow.update { options }
```

Comment:
```suggestion
    fun setOptions(options: ReplayHistorySessionOptions) {
        optionsFlow.value = options
    }
```

Reply:
what's the reason for preference here? `= optionFlow.update { options }` seems nice to me

Reply:
`update` is used when you need to atomically set the new value which is calculated using the previous one. If you look at its implementation, you will see a loop. The function gets the current value, calculates the new one using the lambda and tries to set it. If this fails (because someone set another value during the calculation) it repeats everything until it succeeds. While `optionsFlow.value = options` simply sets the new value without caring about concurrency, which is exactly what you need. And `update` does more than what's necessary. It is unlikely this will be noticeable, but still it looks like a wrong tool used for the job.

Reply:
Hmm right I only looked into `compareAndSet` but looked past the `while(true)`

I've seen StateFlow drop values when set concurrently. Also was considering this a potentially appropriate place for it because it is a public function. But moving to `optionsFlow.value = options` here